dht.go

```diff
@@ -19,6 +19,9 @@ type DhtAPI interface {
 
 	// FindProviders finds peers in the DHT who can provide a specific value
 	// given a key.
+	//
+	// The returned channel will be closed when the request completes or is
+	// canceled.
 	FindProviders(context.Context, path.Path, ...options.DhtFindProvidersOption) (<-chan peer.AddrInfo, error)
```

Comment:
Why aren't we following the pattern here?

Reply:
In this case, I'm asking IPFS to find me some providers for a key. If the request is canceled, we can just say "we're done" and walk away. That is, a partial response is still a _valid_ response.

This is different from listing files/pins because:

1. The request can be considered "complete" even if it returns a subset of the available providers. On the other hand, if `Ls` returned a subset of files/pins, that would be an incomplete result.
2. Beyond "your arguments are invalid" and "this subsystem isn't ready (not bootstrapped)", there aren't really any useful errors the router could return other than "I can't find what you're looking for".

On the other hand, this brings up a good point. IMO, the `Search` function should have the same interface. That is `NameAPI.Search(ctx, name, opts) (<-chan path.Path, error)`. Thoughts?

cc @alanshaw, I think this answers your question as well.